ClientRuntimes/Java/azure-client-runtime/src/main/java/com/microsoft/azure/PagedList.java

```diff
@@ -0,0 +1,245 @@
+                addAll(nextPage.getItems());
+            } catch (CloudException e) {
+                throw new WebServiceException(e.toString(), e);
+            } catch (IOException e) {
+                throw new DataBindingException(e.getMessage(), e);
+            }
+        }
+    }
+
+    private class Itr implements Iterator<E> {
+        Iterator<E> itemsItr;
+
+        public Itr() {
+            itemsItr = items.iterator();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return itemsItr.hasNext() || nextPageLink != null;
+        }
+
+        @Override
+        public E next() {
+            if (!itemsItr.hasNext()) {
+                if (nextPageLink == null) {
+                    throw new NoSuchElementException();
+                } else {
+                    try {
+                        loadPage(nextPageLink);
+                        itemsItr = items.iterator();
+                    } catch (CloudException e) {
+                        throw new WebServiceException(e.toString(), e);
+                    } catch (IOException e) {
+                        throw new DataBindingException(e.getMessage(), e);
+                    }
+                }
+            }
+            return itemsItr.next();
+        }
+
+        @Override
+        public void remove() {
+            itemsItr.remove();
+        }
+    }
+
+    private class ListItr extends Itr implements ListIterator<E> {
+        ListIterator<E> itemsListItr;
+
+        public ListItr(int index) {
+            itemsListItr = items.listIterator(index);
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            return itemsListItr.hasPrevious();
+        }
+
+        @Override
+        public E previous() {
+            return itemsListItr.previous();
+        }
+
+        @Override
+        public int nextIndex() {
+            return itemsListItr.nextIndex();
+        }
+
+        @Override
+        public int previousIndex() {
+            return itemsListItr.previousIndex();
+        }
+
+        @Override
+        public void set(E e) {
+            itemsListItr.set(e);
+        }
+
+        @Override
+        public void add(E e) {
+            itemsListItr.add(e);
+        }
+    }
+
+    @Override
+    public int size() {
+        loadAll();
+        return items.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return items.isEmpty() && nextPageLink == null;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        loadAll();
+        return items.contains(o);
+    }
+
+    @Override
+    public Iterator<E> iterator() {
+        return new Itr();
+    }
+
+    @Override
+    public Object[] toArray() {
+        loadAll();
+        return items.toArray();
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        loadAll();
+        return items.toArray(a);
+    }
+
+    @Override
+    public boolean add(E e) {
+        return items.add(e);
+    }
+
```

Comment:
Fine, thanks for confirming get behavior. - you are right - modifying returned list is not a common scenario for us. 

LGTM